src/proto/training_algorithm.proto

```diff
@@ -26,9 +26,73 @@
 
 syntax = "proto3";
 
+import "google/protobuf/any.proto";
+
 package lbann_data;
 
+// A message describing a training algorithm.
 message TrainingAlgorithm {
-  string type = 1;
-  string name = 3;
+  // Human-decided name. Just for output/user reference.
+  string name = 1;
+
+  // This is a message containing the parameters needed to construct a
+  // solver. The actual message type must be registered with LBANN's
+  // factory. The onus is on the user to ensure the parameters
+  // provided are consistent with the algorithm they have selected.
+  google.protobuf.Any parameters = 2;
 }
+
+// Is-a TrainingAlgorithm
+message SGD {
+  message TerminationCriteria {
+    uint64 max_batches = 1;
+    uint64 max_epochs = 2;
+    double max_seconds = 3;
+  }
+
+  TerminationCriteria stopping_criteria = 1;
+}// message SGD
+
+// Is-a TrainingAlgorithm
+message LTFB {
+  message TerminationCriteria {
+    uint64 max_tournaments = 1;
+  }
+
+  TrainingAlgorithm local_training_algorithm = 1;
+  google.protobuf.Any meta_learning_strategy = 2;
+  TerminationCriteria stopping_criteria = 3;
+}// message LTFB
+
+// The classic LTFB algorithm. Implements MetaLearningStrategy.
+message RandomPairwiseExchange {
+  enum MetricStrategy {
+    LOWER_IS_BETTER = 0;
+    HIGHER_IS_BETTER = 1;
+  }
+
+  string metric_name = 1;
```

Comment:
Immediate extension of this will be support for multiple evaluation metrics, each metric with its own metric strategy.